### PR TITLE
[SPARK-48880][CORE] Avoid throw NullPointerException if driver plugin fails to initialize

### PR DESCRIPTION
**Pull request by ulysses-you**
### What changes were proposed in this pull request?

This pr skips clear memoryStore if memoryManager is null. This could happen if driver plugin fails to initialize, since we initialize MemoryManager after DriverPlugin.

### Why are the changes needed?

before it would throw:
```
{"class":"java.lang.NullPointerException","msg":"Cannot invoke \"org.apache.spark.memory.MemoryManager.maxOnHeapStorageMemory()\" because \"this.memoryManager\" is null","stacktrace":[{"class":"org.apache.spark.storage.memory.MemoryStore","method":"maxMemory","file":"MemoryStore.scala","line":110},
{"class":"org.apache.spark.storage.memory.MemoryStore","method":"<init>","file":"MemoryStore.scala","line":113},
{"class":"org.apache.spark.storage.BlockManager","method":"memoryStore$lzycompute","file":"BlockManager.scala","line":234},
{"class":"org.apache.spark.storage.BlockManager","method":"memoryStore","file":"BlockManager.scala","line":233},
{"class":"org.apache.spark.storage.BlockManager","method":"stop","file":"BlockManager.scala","line":2167},
{"class":"org.apache.spark.SparkEnv","method":"stop","file":"SparkEnv.scala","line":118},
{"class":"org.apache.spark.SparkContext","method":"$anonfun$stop$25","file":"SparkContext.scala","line":2369},
{"class":"org.apache.spark.util.Utils$","method":"tryLogNonFatalError","file":"Utils.scala","line":1299},
{"class":"org.apache.spark.SparkContext","method":"stop","file":"SparkContext.scala","line":2369}
```

### Does this PR introduce _any_ user-facing change?

no

### How was this patch tested?

manually test

### Was this patch authored or co-authored using generative AI tooling?

no

### CODE REVIEW

**Comment by ulysses-you** [comment]
cc @yaooqinn @cloud-fan thank you

**Comment by yaooqinn** [comment]
Is it more likely that the issue is due to the driver plugin coupled with spark internal classes and its failing?

**Comment by ulysses-you** [comment]
@yaooqinn it does not require touch spark internal classes in driver plugin. This error happened if we start spark-shell  using `SparkConnectPlugin` but failed to bind address(The 15002 port has been used).

**Review by yaooqinn — APPROVED** [review]

**Comment by ulysses-you** [comment]
thanks, merging to master

File: core/src/main/scala/org/apache/spark/storage/BlockManager.scala [hunk 1]
```diff
@@ -2164,7 +2164,10 @@ private[spark] class BlockManager(
     diskBlockManager.stop()
     rpcEnv.stop(storageEndpoint)
     blockInfoManager.clear()
-    memoryStore.clear()
+    // The memoryManager may be null if the driver plugin fails to initialize
```
Comment (cloud-fan) [inline core/src/main/scala/org/apache/spark/storage/BlockManager.scala:2167]: Should we fix it the other way around? Do not break Spark initialization if the plugin fails to initialize.

Comment (ulysses-you) [reply core/src/main/scala/org/apache/spark/storage/BlockManager.scala:2167]: I'm not sure if we should keep starting spark context if driver plguin fail to initialize. Before we have fixed an issue for executor plugin [SPARK-40320](https://issues.apache.org/jira/browse/SPARK-40320) , by making exexutor fast fail when it fails to initialize. So this pr leave driver fast fail and just drops the noisy error message.

Comment (cloud-fan) [reply core/src/main/scala/org/apache/spark/storage/BlockManager.scala:2167]: I see

Comment (mridulm) [reply core/src/main/scala/org/apache/spark/storage/BlockManager.scala:2167]: Agree, if driver plugin fails to initialize, we should fail fast - the functionality could (and tends to be) critical for the spark application: for example, remote shuffle service initialization.

**Pull request closed without merging**